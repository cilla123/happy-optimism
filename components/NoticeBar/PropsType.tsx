import MessagePropsType from '../Message/PropsType';

export default interface PropsType extends MessagePropsType {
  autoscroll?: boolean;
}
