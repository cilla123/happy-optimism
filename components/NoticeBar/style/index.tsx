import '../../style/';
import '../../Message/style/';
import '../../Icon/style/';
import './index.scss';
