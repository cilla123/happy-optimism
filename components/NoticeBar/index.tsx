import React, { PureComponent } from 'react';
import PropsType from './PropsType';
import Message from '../Message';
import Icon from '../Icon';

export interface MessageProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class NoticeBar extends PureComponent<MessageProps, any> {

  static defaultProps = {
    prefixCls: 'za-noticebar',
    theme: 'warning',
    icon: <Icon type="broadcast" />,
    hasArrow: false,
    hasClosable: false,
    autoscroll: false,
  };

  private wrapper;
  private content;
  private moveInterval;

  constructor(props) {
    super(props);
    this.state = {
      offset: 0,
    };
  }

  componentDidMount() {
    const { autoscroll } = this.props;
    if (!autoscroll) {
      return;
    }

    const distance = this.wrapper.offsetWidth - this.content.offsetWidth;
    if (distance > 0) {
      return;
    }

    let delay = 1000;
    this.moveInterval = setInterval(() => {
      let { offset } = this.state;
      if ((offset < distance || offset >= 0) && delay > 0) {
        delay -= 50;
        return;
      }

      delay = 1000;
      offset = (offset < distance)
        ? 0
        : (offset - 1);

      this.setState({ offset });
    }, 50);
  }

  componentWillUnmount() {
    if (this.moveInterval) {
      clearInterval(this.moveInterval);
    }
  }

  render() {
    const { prefixCls, children, autoscroll, ...others } = this.props;

    return (
      <Message {...others} size="lg">
        <div className={prefixCls} ref={(ele) => { this.wrapper = ele; }}>
          <div
            className={`${prefixCls}-body`}
            ref={(ele) => { this.content = ele; }}
            style={{ left: this.state.offset }}
          >
            {children}
          </div>
        </div>
      </Message>
    );
  }
}
