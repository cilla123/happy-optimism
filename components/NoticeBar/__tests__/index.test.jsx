import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import NoticeBar from '../index';

describe('NoticeBar', () => {
  it('renders correctly', () => {
    jest.useFakeTimers();
    const wrapper = mount(<NoticeBar>foo</NoticeBar>);
    jest.runTimersToTime(3000);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('theme', () => {
    const wrapper = render(<NoticeBar theme="error">foo</NoticeBar>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('icon', () => {
    const wrapper = render(<NoticeBar icon={<img alt="" src="https://zhongantecheng.github.io/zarm/images/state.18e78939.png" />}>foo</NoticeBar>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('autoscroll', () => {
    jest.useFakeTimers();
    const wrapper = mount(<NoticeBar autoscroll>各位zarmer请注意，本组件使用了自动滚动功能，更多用法请参见使用文档。</NoticeBar>);
    jest.runTimersToTime(3000);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
