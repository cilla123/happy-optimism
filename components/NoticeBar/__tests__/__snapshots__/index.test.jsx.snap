// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoticeBar autoscroll 1`] = `
<NoticeBar
  autoscroll={true}
  hasArrow={false}
  hasClosable={false}
  icon={
    <Icon
      prefixCls="za-icon"
      type="broadcast"
    />
  }
  prefixCls="za-noticebar"
  theme="warning"
>
  <Message
    hasArrow={false}
    hasClosable={false}
    icon={
      <Icon
        prefixCls="za-icon"
        type="broadcast"
      />
    }
    prefixCls="za-message"
    size="lg"
    theme="warning"
  >
    <div
      className="za-message theme-warning size-lg"
      onClick={[Function]}
    >
      <div
        className="za-message-header"
      >
        <div
          className="za-message-icon"
        >
          <Icon
            prefixCls="za-icon"
            type="broadcast"
          >
            <i
              className="za-icon za-icon-broadcast"
            />
          </Icon>
        </div>
      </div>
      <div
        className="za-message-body"
      >
        <div
          className="za-noticebar"
        >
          <div
            className="za-noticebar-body"
            style={
              Object {
                "left": 0,
              }
            }
          >
            各位zarmer请注意，本组件使用了自动滚动功能，更多用法请参见使用文档。
          </div>
        </div>
      </div>
    </div>
  </Message>
</NoticeBar>
`;

exports[`NoticeBar icon 1`] = `
<div
  class="za-message theme-warning size-lg"
>
  <div
    class="za-message-header"
  >
    <div
      class="za-message-icon"
    >
      <img
        alt=""
        src="https://zhongantecheng.github.io/zarm/images/state.18e78939.png"
      />
    </div>
  </div>
  <div
    class="za-message-body"
  >
    <div
      class="za-noticebar"
    >
      <div
        class="za-noticebar-body"
        style="left:0"
      >
        foo
      </div>
    </div>
  </div>
</div>
`;

exports[`NoticeBar renders correctly 1`] = `
<NoticeBar
  autoscroll={false}
  hasArrow={false}
  hasClosable={false}
  icon={
    <Icon
      prefixCls="za-icon"
      type="broadcast"
    />
  }
  prefixCls="za-noticebar"
  theme="warning"
>
  <Message
    hasArrow={false}
    hasClosable={false}
    icon={
      <Icon
        prefixCls="za-icon"
        type="broadcast"
      />
    }
    prefixCls="za-message"
    size="lg"
    theme="warning"
  >
    <div
      className="za-message theme-warning size-lg"
      onClick={[Function]}
    >
      <div
        className="za-message-header"
      >
        <div
          className="za-message-icon"
        >
          <Icon
            prefixCls="za-icon"
            type="broadcast"
          >
            <i
              className="za-icon za-icon-broadcast"
            />
          </Icon>
        </div>
      </div>
      <div
        className="za-message-body"
      >
        <div
          className="za-noticebar"
        >
          <div
            className="za-noticebar-body"
            style={
              Object {
                "left": 0,
              }
            }
          >
            foo
          </div>
        </div>
      </div>
    </div>
  </Message>
</NoticeBar>
`;

exports[`NoticeBar theme 1`] = `
<div
  class="za-message theme-error size-lg"
>
  <div
    class="za-message-header"
  >
    <div
      class="za-message-icon"
    >
      <i
        class="za-icon za-icon-broadcast"
      />
    </div>
  </div>
  <div
    class="za-message-body"
  >
    <div
      class="za-noticebar"
    >
      <div
        class="za-noticebar-body"
        style="left:0"
      >
        foo
      </div>
    </div>
  </div>
</div>
`;
